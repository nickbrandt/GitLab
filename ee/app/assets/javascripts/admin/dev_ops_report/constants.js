import { s__, __, sprintf } from '~/locale';

export const MAX_SEGMENTS = 30;

export const MAX_REQUEST_COUNT = 10;

export const DEVOPS_ADOPTION_SEGMENT_MODAL_ID = 'devopsSegmentModal';

export const DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID = 'devopsSegmentDeleteModal';

export const DATE_TIME_FORMAT = 'yyyy-mm-dd HH:MM';

export const DEVOPS_ADOPTION_ERROR_KEYS = {
  groups: 'groupsError',
  segments: 'segmentsError',
};

export const DEVOPS_ADOPTION_STRINGS = {
  app: {
    [DEVOPS_ADOPTION_ERROR_KEYS.groups]: s__(
      'DevopsAdoption|There was an error fetching Groups. Please refresh the page to try again.',
    ),
    [DEVOPS_ADOPTION_ERROR_KEYS.segments]: s__(
      'DevopsAdoption|There was an error fetching Group adoption data. Please refresh the page to try again.',
    ),
    tableHeader: {
      text: s__(
        'DevopsAdoption|Feature adoption is based on usage in the last calendar month. Last updated: %{timestamp}.',
      ),
      button: s__('DevopsAdoption|Add Group'),
      buttonTooltip: sprintf(s__('DevopsAdoption|Maximum %{maxSegments} groups allowed'), {
        maxSegments: MAX_SEGMENTS,
      }),
    },
  },
  emptyState: {
    title: s__('DevopsAdoption|Add a group to get started'),
    description: s__(
      'DevopsAdoption|DevOps adoption tracks the use of key features across your favorite groups. Add a group to the table to begin.',
    ),
    button: s__('DevopsAdoption|Add Group'),
  },
  modal: {
    addingTitle: s__('DevopsAdoption|Add Group'),
    addingButton: s__('DevopsAdoption|Add Group'),
    editingButton: s__('DevopsAdoption|Save changes'),
    cancel: __('Cancel'),
    namePlaceholder: s__('DevopsAdoption|My group'),
    filterPlaceholder: s__('DevopsAdoption|Filter by name'),
    selectedGroupsTextSingular: s__('DevopsAdoption|%{selectedCount} group selected'),
    selectedGroupsTextPlural: s__('DevopsAdoption|%{selectedCount} groups selected'),
    error: s__('DevopsAdoption|An error occured while saving the group. Please try again.'),
    noResults: s__('DevopsAdoption|No filter results.'),
  },
  table: {
    removeButton: s__('DevopsAdoption|Remove Group from the table'),
    headers: {
      name: {
        label: __('Group'),
      },
      issueOpened: {
        label: s__('DevopsAdoption|Issues'),
        tooltip: s__('DevopsAdoption|At least 1 issue opened'),
      },
      mergeRequestOpened: {
        label: s__('DevopsAdoption|MRs'),
        tooltip: s__('DevopsAdoption|At least 1 MR opened'),
      },
      mergeRequestApproved: {
        label: s__('DevopsAdoption|Approvals'),
        tooltip: s__('DevopsAdoption|At least 1 approval on an MR'),
      },
      runnerConfigured: {
        label: s__('DevopsAdoption|Runners'),
        tooltip: s__('DevopsAdoption|Runner configured for project/group'),
      },
      pipelineSucceeded: {
        label: s__('DevopsAdoption|Pipelines'),
        tooltip: s__('DevopsAdoption|At least 1 pipeline successfully run'),
      },
      deploySucceeded: {
        label: s__('DevopsAdoption|Deploys'),
        tooltip: s__('DevopsAdoption|At least 1 deploy'),
      },
      securityScanSucceeded: {
        label: s__('DevopsAdoption|Scanning'),
        tooltip: s__('DevopsAdoption|At least 1 security scan of any type run in pipeline'),
      },
    },
  },
  deleteModal: {
    title: s__('DevopsAdoption|Confirm remove Group'),
    confirmationMessage: s__(
      'DevopsAdoption|Are you sure that you would like to remove %{name} from the table?',
    ),
    cancel: __('Cancel'),
    confirm: s__('DevopsAdoption|Remove Group'),
    error: s__('DevopsAdoption|An error occured while removing the group. Please try again.'),
  },
  tableCell: {
    trueText: s__('DevopsAdoption|Adopted'),
    falseText: s__('DevopsAdoption|Not adopted'),
  },
};

export const DEVOPS_ADOPTION_TABLE_TEST_IDS = {
  TABLE_HEADERS: 'header',
  SEGMENT: 'segmentCol',
  ISSUES: 'issuesCol',
  MRS: 'mrsCol',
  APPROVALS: 'approvalsCol',
  RUNNERS: 'runnersCol',
  PIPELINES: 'pipelinesCol',
  DEPLOYS: 'deploysCol',
  ACTIONS: 'actionsCol',
  SCANNING: 'scanningCol',
  LOCAL_STORAGE_SORT_BY: 'localStorageSortBy',
  LOCAL_STORAGE_SORT_DESC: 'localStorageSortDesc',
};

export const DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_BY_STORAGE_KEY =
  'devops_adoption_segments_table_sort_by';

export const DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_DESC_STORAGE_KEY =
  'devops_adoption_segments_table_sort_desc';
