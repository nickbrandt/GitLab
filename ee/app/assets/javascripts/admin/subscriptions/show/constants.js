import { __, s__ } from '~/locale';
import activateSubscriptionMutation from './graphql/mutations/activate_subscription.mutation.graphql';
import getCurrentLicense from './graphql/queries/get_current_license.query.graphql';
import getLicenseHistory from './graphql/queries/get_license_history.query.graphql';

export const fieldRequiredMessage = s__('SuperSonics|This field is required.');
export const subscriptionMainTitle = s__('SuperSonics|Your subscription');
export const subscriptionActivationNotificationText = s__(
  `SuperSonics|Your subscription was successfully activated. You can see the details below.`,
);
export const subscriptionActivationInsertCode = __(
  "If you've purchased or renewed your subscription and have an activation code, please enter it below to start the activation process.",
);
export const howToActivateSubscription = s__(
  'SuperSonics|Learn how to %{linkStart}activate your subscription%{linkEnd}.',
);
export const activateLabel = s__('SuperSonics|Activate');
export const activateSubscription = s__('SuperSonics|Activate subscription');
export const activateCloudLicense = s__('SuperSonics|Activate cloud license');
export const noActiveSubscription = s__(`SuperSonics|You do not have an active subscription`);
export const subscriptionDetailsHeaderText = s__('SuperSonics|Subscription details');
export const licensedToHeaderText = s__('SuperSonics|Licensed to');
export const manageSubscriptionButtonText = s__('SuperSonics|Manage');
export const syncSubscriptionButtonText = s__('SuperSonics|Sync subscription details');
export const copySubscriptionIdButtonText = __('Copy');
export const licenseFileText = __('License file');
export const cloudLicenseText = s__('SuperSonics|Cloud license');
export const usersInSubscriptionUnlimited = __('Unlimited');
export const detailsLabels = {
  address: __('Address'),
  company: __('Company'),
  email: __('Email'),
  id: s__('SuperSonics|ID'),
  lastSync: s__('SuperSonics|Last Sync'),
  name: __('Name'),
  plan: s__('SuperSonics|Plan'),
  expiresAt: s__('SuperSonics|Renews'),
  startsAt: s__('SuperSonics|Started'),
};

export const removeLicense = __('Remove license');
export const removeLicenseConfirm = __('Are you sure you want to remove the license?');
export const uploadLicense = __('Upload license');
export const uploadLicenseFile = s__('SuperSonics|Upload a license file');
export const billableUsersTitle = s__('SuperSonics|Billable users');
export const maximumUsersTitle = s__('SuperSonics|Maximum users');
export const usersInSubscriptionTitle = s__('SuperSonics|Users in subscription');
export const usersOverSubscriptionTitle = s__('SuperSonics|Users over subscription');
export const billableUsersText = s__(
  'SuperSonics|This is the number of %{billableUsersLinkStart}billable users%{billableUsersLinkEnd} on your installation, and this is the minimum number you need to purchase when you renew your license.',
);
export const maximumUsersText = s__(
  'SuperSonics|This is the highest peak of users on your installation since the license started.',
);
export const usersInSubscriptionText = s__(
  `SuperSonics|Users with a Guest role or those who don't belong to a Project or Group will not use a seat from your license.`,
);
export const usersOverSubscriptionText = s__(
  `SuperSonics|You'll be charged for %{trueUpLinkStart}users over license%{trueUpLinkEnd} on a quarterly or annual basis, depending on the terms of your agreement.`,
);
export const subscriptionTable = {
  activatedAt: s__('SuperSonics|Activated on'),
  expiresOn: s__('SuperSonics|Expires on'),
  seats: s__('SuperSonics|Seats'),
  startsAt: s__('SuperSonics|Valid From'),
  title: __('Subscription History'),
  type: s__('SuperSonics|Type'),
};
export const connectivityIssue = s__('SuperSonics|There is a connectivity issue.');
export const manualSyncSuccessfulTitle = s__(
  'SuperSonics|The subscription details synced successfully.',
);
export const manualSyncFailureText = s__(
  'SuperSonics|You can no longer sync your subscription details with GitLab. Get help for the most common connectivity issues by %{connectivityHelpLinkStart}troubleshooting the activation code%{connectivityHelpLinkEnd}.',
);

export const subscriptionActivationForm = {
  activationCode: s__('SuperSonics|Activation code'),
  pasteActivationCode: s__('SuperSonics|Paste your activation code'),
  acceptTerms: s__(
    'SuperSonics|I agree that my use of the GitLab Software is subject to the Subscription Agreement located at the %{linkStart}Terms of Service%{linkEnd}, unless otherwise agreed to in writing with GitLab.',
  ),
};

export const subscriptionSyncStatus = {
  SYNC_FAILURE: 'SYNC_FAILURE',
  SYNC_SUCCESS: 'SYNC_SUCCESS',
};
export const subscriptionTypes = {
  CLOUD: 'cloud',
  LICENSE_FILE: 'license_file',
};

export const subscriptionQueries = {
  query: getCurrentLicense,
  mutation: activateSubscriptionMutation,
};

export const subscriptionHistoryQueries = {
  query: getLicenseHistory,
};

export const trialCard = {
  title: s__('SuperSonics|Free trial'),
  description: s__(
    'SuperSonics|You can start a free trial of GitLab Ultimate without any obligation or payment details.',
  ),
  startTrial: s__('SuperSonics|Start free trial'),
};

export const buySubscriptionCard = {
  title: s__('SuperSonics|Subscription'),
  description: s__(
    'SuperSonics|Ready to get started? A GitLab plan is ideal for scaling organizations and for multi team usage.',
  ),
  buttonLabel: s__('SuperSonics|Buy subscription'),
};

export const INVALID_CODE_ERROR_MESSAGE = 'invalid activation code';
export const CONNECTIVITY_ERROR = 'CONNECTIVITY_ERROR';
export const INVALID_CODE_ERROR = 'INVALID_CODE_ERROR';
export const generalActivationErrorTitle = s__(
  'SuperSonics|An error occurred while activating your subscription.',
);
export const generalActivationErrorMessage = s__(
  'SuperSonics|You can learn more about %{activationLinkStart}activating your subscription%{activationLinkEnd}. If you need further assistance, please %{supportLinkStart}contact GitLab Support%{supportLinkEnd}.',
);
export const invalidActivationCode = s__(
  'SuperSonics|The activation code is not valid. Please make sure to copy it exactly from the Customers Portal or confirmation email. Learn more about %{linkStart}activating your subscription%{linkEnd}.',
);
export const connectivityErrorAlert = {
  subtitle: s__(
    'SuperSonics|To activate your subscription, connect to GitLab servers through the %{linkStart}Cloud Licensing%{linkEnd} service, a hassle-free way to manage your subscription.',
  ),
  helpText: s__(
    'SuperSonics|Get help for the most common connectivity issues by %{linkStart}troubleshooting the activation code%{linkEnd}.',
  ),
};
export const supportLink = 'https://about.gitlab.com/support/#contact-support';
