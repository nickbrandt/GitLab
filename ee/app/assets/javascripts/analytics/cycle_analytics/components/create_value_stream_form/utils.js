import { isEqual, pick } from 'lodash';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { isStartEvent, getAllowedEndEvents, eventToOption, eventsByIdentifier } from '../../utils';
import {
  i18n,
  ERRORS,
  defaultErrors,
  defaultFields,
  NAME_MAX_LENGTH,
  formFieldKeys,
} from './constants';

/**
 * @typedef {Object} CustomStageEvents
 * @property {String} canBeStartEvent - Title of the metric measured
 * @property {String} name - Friendly name for the event
 * @property {String} identifier - snakeized name for the event
 *
 * @typedef {Object} DropdownData
 * @property {String} text - Friendly name for the event
 * @property {String} value - Value to be submitted for the dropdown
 */

/**
 * Takes an array of custom stage events to return only the
 * events where `canBeStartEvent` is true and converts them
 * to { value, text } pairs for use in dropdowns
 *
 * @param {CustomStageEvents[]} events
 * @returns {DropdownData[]} array of start events formatted for dropdowns
 */
export const startEventOptions = (eventsList) => [
  { value: null, text: i18n.SELECT_START_EVENT },
  ...eventsList.filter(isStartEvent).map(eventToOption),
];

/**
 * Takes an array of custom stage events to return only the
 * events where `canBeStartEvent` is false and converts them
 * to { value, text } pairs for use in dropdowns
 *
 * @param {CustomStageEvents[]} events
 * @returns {DropdownData[]} array end events formatted for dropdowns
 */
export const endEventOptions = (eventsList, startEventIdentifier) => {
  const endEvents = getAllowedEndEvents(eventsList, startEventIdentifier);
  return [
    { value: null, text: i18n.SELECT_END_EVENT },
    ...eventsByIdentifier(eventsList, endEvents).map(eventToOption),
  ];
};

/**
 * @typedef {Object} CustomStageFormData
 * @property {Object.<String, String>} fields - form field values
 * @property {Object.<String, Array>} errors - form field errors
 */

/**
 * Initializes the fields and errors for the custom stages form
 * providing defaults for any missing keys
 *
 * @param {CustomStageFormData} data
 * @returns {CustomStageFormData} the updated initial data with all defaults
 */
export const initializeFormData = ({ fields, errors }) => {
  const initErrors = fields?.endEventIdentifier
    ? defaultErrors
    : {
        ...defaultErrors,
        endEventIdentifier: !fields?.startEventIdentifier ? [ERRORS.START_EVENT_REQUIRED] : [],
      };
  return {
    fields: {
      ...defaultFields,
      ...fields,
    },
    errors: {
      ...initErrors,
      ...errors,
    },
  };
};

/**
 * Validates the form fields for the custom stages form
 * Any errors will be returned in a object where the key is
 * the name of the field.g
 *
 * @param {Object} fields key value pair of form field values
 * @param {Object} defaultStageNames array of lower case default value stream names
 * @returns {Object} key value pair of form fields with an array of errors
 */
export const validateStage = (fields, defaultStageNames = []) => {
  const newErrors = {};

  if (fields?.name) {
    if (fields.name.length > NAME_MAX_LENGTH) {
      newErrors.name = [ERRORS.MAX_LENGTH];
    }
    if (fields?.custom && defaultStageNames.includes(fields.name.toLowerCase())) {
      newErrors.name = [ERRORS.STAGE_NAME_EXISTS];
    }
  } else {
    newErrors.name = [ERRORS.STAGE_NAME_MIN_LENGTH];
  }

  if (fields?.startEventIdentifier) {
    if (!fields?.endEventIdentifier) {
      newErrors.endEventIdentifier = [ERRORS.END_EVENT_REQUIRED];
    }
  } else {
    newErrors.endEventIdentifier = [ERRORS.START_EVENT_REQUIRED];
  }
  return newErrors;
};

/**
 * Validates the name of a value stream Any errors will be
 * returned as an array in a object with key`name`
 *
 * @param {Object} fields key value pair of form field values
 * @returns {Array} an array of errors
 */
export const validateValueStreamName = ({ name = '' }) => {
  const errors = [];
  if (name.length > NAME_MAX_LENGTH) {
    errors.push(ERRORS.MAX_LENGTH);
  }
  if (!name.length) {
    errors.push(ERRORS.VALUE_STREAM_NAME_MIN_LENGTH);
  }
  return errors;
};

/**
 * Formats the value stream stages for submission, ensures that the
 * 'custom' property is set when we are editing, we include the `id` if its
 * set and all fields are converted to snake case
 *
 * @param {Array} stages array of value stream stages
 * @param {Boolean} isEditing flag to indicate if we are editing a value stream or creating
 * @returns {Array} the array prepared to be submitted for persistence
 */
export const formatStageDataForSubmission = (stages, isEditing = false) => {
  return stages.map(({ id = null, custom = false, name, ...rest }) => {
    let editProps = { custom };
    if (isEditing) {
      // We can add a new stage to the value stream when either creating, or editing
      // If a new stage has been added then at this point, the `id` won't exist
      // The new stage is still `custom` but wont have an id until the form submits and its persisted to the DB
      editProps = id ? { id, custom: true } : { custom: true };
    }
    // While we work on https://gitlab.com/gitlab-org/gitlab/-/issues/321959 we should not allow editing default
    return custom
      ? convertObjectPropsToSnakeCase({ ...rest, ...editProps, name })
      : convertObjectPropsToSnakeCase({ ...editProps, name, custom: false });
  });
};

/**
 * Checks an array of value stream stages to see if there are
 * any differences in the values they contain
 *
 * @param {Array} stages array of value stream stages
 * @param {Array} stages array of value stream stages
 * @returns {Boolean} returns true if there is a difference in the 2 arrays
 */
export const hasDirtyStage = (currentStages, originalStages) => {
  const cs = currentStages.map((s) => pick(s, formFieldKeys));
  const os = originalStages.map((s) => pick(s, formFieldKeys));
  return !isEqual(cs, os);
};

/**
 * Checks if the target name matches the name of any of the value
 * stream stages passed in
 *
 * @param {Array} stages array of value stream stages
 * @param {String} targetName name we are searching for
 * @returns {Object} returns the found object or null
 */
const findStageByName = (stages, targetName = '') =>
  stages.find(({ name }) => name.toLowerCase().trim() === targetName.toLowerCase().trim());

/**
 * Returns a valid custom value stream stage
 *
 * @param {Object} stage a raw value stream stage retrieved from the vuex store
 * @returns {Object} the same stage with fields adjusted for the value stream form
 */
const prepareCustomStage = ({ startEventLabel = {}, endEventLabel = {}, ...rest }) => ({
  ...rest,
  startEventLabel,
  endEventLabel,
  startEventLabelId: startEventLabel?.id || null,
  endEventLabelId: endEventLabel?.id || null,
  isDefault: false,
});

/**
 * Returns a valid default value stream stage
 *
 * @param {Object} stage a raw value stream stage retrieved from the vuex store
 * @returns {Object} the same stage with fields adjusted for the value stream form
 */
const prepareDefaultStage = (defaultStageConfig, { name, ...rest }) => {
  // default stages currently dont have any label based events
  const stage = findStageByName(defaultStageConfig, name) || null;
  if (!stage) return {};
  const { startEventIdentifier = null, endEventIdentifier = null } = stage;
  return {
    ...rest,
    name,
    startEventIdentifier,
    endEventIdentifier,
    isDefault: true,
  };
};

const generateHiddenDefaultStages = (defaultStageConfig, stageNames) => {
  // We use the stage name to check for any default stages that might be hidden
  // Currently the default stages can't be renamed
  return defaultStageConfig
    .filter(({ name }) => !stageNames.includes(name.toLowerCase()))
    .map((data) => ({ ...data, hidden: true }));
};

/**
 * Returns a valid array of value stream stages for
 * use in the value stream form
 *
 * @param {Array} defaultStageConfig an array of the default value stream stages retrieved from the backend
 * @param {Array} selectedValueStreamStages an array of raw value stream stages retrieved from the vuex store
 * @returns {Object} the same stage with fields adjusted for the value stream form
 */
export const generateInitialStageData = (defaultStageConfig, selectedValueStreamStages) => {
  const hiddenDefaultStages = generateHiddenDefaultStages(
    defaultStageConfig,
    selectedValueStreamStages.map((s) => s.name.toLowerCase()),
  );
  const combinedStages = [...selectedValueStreamStages, ...hiddenDefaultStages];
  return combinedStages.map(
    ({ startEventIdentifier = null, endEventIdentifier = null, custom = false, ...rest }) => {
      const stageData =
        custom && startEventIdentifier && endEventIdentifier
          ? prepareCustomStage({ ...rest, startEventIdentifier, endEventIdentifier })
          : prepareDefaultStage(defaultStageConfig, rest);

      if (stageData?.name) {
        return {
          ...stageData,
          custom,
        };
      }
      return {};
    },
  );
};
