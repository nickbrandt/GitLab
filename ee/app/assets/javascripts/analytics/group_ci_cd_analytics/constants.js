export const STAT_ERROR_PLACEHOLDER = '-';
