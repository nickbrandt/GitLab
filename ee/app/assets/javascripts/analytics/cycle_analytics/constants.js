import { OVERVIEW_STAGE_ID } from '~/cycle_analytics/constants';
import { __, s__ } from '~/locale';

export const EVENTS_LIST_ITEM_LIMIT = 50;

export const TASKS_BY_TYPE_SUBJECT_ISSUE = 'Issue';
export const TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST = 'MergeRequest';
export const TASKS_BY_TYPE_MAX_LABELS = 15;

export const TASKS_BY_TYPE_SUBJECT_FILTER_OPTIONS = {
  [TASKS_BY_TYPE_SUBJECT_ISSUE]: __('Issues'),
  [TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST]: __('Merge Requests'),
};

export const TASKS_BY_TYPE_FILTERS = {
  SUBJECT: 'SUBJECT',
  LABEL: 'LABEL',
};

export const DEFAULT_VALUE_STREAM_ID = 'default';

export const OVERVIEW_METRICS = {
  TIME_SUMMARY: 'TIME_SUMMARY',
  RECENT_ACTIVITY: 'RECENT_ACTIVITY',
};

export const FETCH_VALUE_STREAM_DATA = 'fetchValueStreamData';

export const OVERVIEW_STAGE_CONFIG = {
  id: OVERVIEW_STAGE_ID,
  slug: OVERVIEW_STAGE_ID,
  title: __('Overview'),
  icon: 'home',
};

export const NOT_ENOUGH_DATA_ERROR = s__(
  "ValueStreamAnalyticsStage|We don't have enough data to show this stage.",
);

export const PAGINATION_TYPE = 'keyset';
export const PAGINATION_SORT_FIELD_END_EVENT = 'end_event';
export const PAGINATION_SORT_FIELD_DURATION = 'duration';
export const PAGINATION_SORT_DIRECTION_DESC = 'desc';
export const PAGINATION_SORT_DIRECTION_ASC = 'asc';

export const METRICS_POPOVER_CONTENT = {
  'lead-time': {
    description: s__('ValueStreamAnalytics|Median time from issue created to issue closed.'),
  },
  'cycle-time': {
    description: s__(
      'ValueStreamAnalytics|Median time from issue first merge request created to issue closed.',
    ),
  },
  'new-issues': { description: s__('ValueStreamAnalytics|Number of new issues created.') },
  deploys: { description: s__('ValueStreamAnalytics|Total number of deploys to production.') },
  'deployment-frequency': {
    description: s__('ValueStreamAnalytics|Average number of deployments to production per day.'),
  },
};
