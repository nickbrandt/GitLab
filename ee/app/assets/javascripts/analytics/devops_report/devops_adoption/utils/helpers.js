import { isToday } from '~/lib/utils/datetime_utility';

/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, timestamp and check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, timestamp, openModal }) => {
  if (openModal) {
    return false;
  } else if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );
  const dataNotRefreshedToday = !isToday(new Date(timestamp));

  return anyPendingEnabledNamespaces || dataNotRefreshedToday;
};
