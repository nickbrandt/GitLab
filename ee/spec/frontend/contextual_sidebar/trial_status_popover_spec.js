import { GlPopover } from '@gitlab/ui';
import { GlBreakpointInstance } from '@gitlab/ui/dist/utils';
import { mount, shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import { POPOVER, TRACKING_PROPERTY } from 'ee/contextual_sidebar/components/constants';
import TrialStatusPopover from 'ee/contextual_sidebar/components/trial_status_popover.vue';
import { mockTracking, unmockTracking } from 'helpers/tracking_helper';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';

Vue.config.ignoredElements = ['gl-emoji'];

describe('TrialStatusPopover component', () => {
  let wrapper;
  let trackingSpy;

  const { trackingEvents } = POPOVER;

  const findGlPopover = () => wrapper.findComponent(GlPopover);

  const expectTracking = ({ action, ...options } = {}) => {
    return expect(trackingSpy).toHaveBeenCalledWith(undefined, action, {
      ...options,
      property: TRACKING_PROPERTY,
    });
  };

  const createComponent = (providers = {}, mountFn = shallowMount) => {
    return extendedWrapper(
      mountFn(TrialStatusPopover, {
        provide: {
          groupName: 'Some Test Group',
          planName: 'Ultimate',
          plansHref: 'billing/path-for/group',
          purchaseHref: 'transactions/new',
          targetId: 'target-element-identifier',
          trialEndDate: new Date('2021-02-28'),
          ...providers,
        },
      }),
    );
  };

  beforeEach(() => {
    wrapper = createComponent();
    trackingSpy = mockTracking('_category_', wrapper.element, jest.spyOn);
  });

  afterEach(() => {
    wrapper.destroy();
    unmockTracking();
  });

  describe('interpolated strings', () => {
    it('correctly interpolates them all', () => {
      wrapper = createComponent(mount);

      expect(wrapper.text()).not.toMatch(/%{\w+}/);
    });
  });

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('tracks when the upgrade button is clicked', () => {
    wrapper.findByTestId('upgradeBtn').vm.$emit('click');

    expectTracking(trackingEvents.upgradeBtnClick);
  });

  it('tracks when the compare button is clicked', () => {
    wrapper.findByTestId('compareBtn').vm.$emit('click');

    expectTracking(trackingEvents.compareBtnClick);
  });

  describe('startInitiallyShown', () => {
    describe('when set to true', () => {
      beforeEach(() => {
        wrapper = createComponent({ startInitiallyShown: true });
      });

      it('causes the popover to be shown by default', () => {
        expect(findGlPopover().attributes('show')).toBeTruthy();
      });

      it('removes the popover triggers', () => {
        expect(findGlPopover().attributes('triggers')).toBe('');
      });
    });

    describe('when set to false', () => {
      beforeEach(() => {
        wrapper = createComponent({ startInitiallyShown: false });
      });

      it('does not cause the popover to be shown by default', () => {
        expect(findGlPopover().attributes('show')).toBeFalsy();
      });

      it('uses the standard triggers for the popover', () => {
        expect(findGlPopover().attributes('triggers')).toBe('hover focus');
      });
    });
  });

  describe('close button', () => {
    describe('when the popover starts off forcibly shown', () => {
      beforeEach(() => {
        wrapper = createComponent({ startInitiallyShown: true }, mount);
      });

      it('is rendered', () => {
        expect(wrapper.findByTestId('closeBtn').exists()).toBeTruthy();
      });

      describe('when clicked', () => {
        beforeEach(async () => {
          wrapper.findByTestId('closeBtn').trigger('click');
          await wrapper.vm.$nextTick();
        });

        it('closes the popover component', () => {
          expect(findGlPopover().props('show')).toBeFalsy();
        });

        it('tracks an event', () => {
          expectTracking(trackingEvents.closeBtnClick);
        });

        it('continues to be shown in the popover', () => {
          expect(wrapper.findByTestId('closeBtn').exists()).toBeTruthy();
        });
      });
    });

    describe('when the popover does not start off forcibly shown', () => {
      it('is not rendered', () => {
        expect(wrapper.findByTestId('closeBtn').exists()).toBeFalsy();
      });
    });
  });

  describe('methods', () => {
    describe('onResize', () => {
      it.each`
        bp      | isDisabled
        ${'xs'} | ${'true'}
        ${'sm'} | ${'true'}
        ${'md'} | ${undefined}
        ${'lg'} | ${undefined}
        ${'xl'} | ${undefined}
      `(
        'sets disabled to `$isDisabled` when the breakpoint is "$bp"',
        async ({ bp, isDisabled }) => {
          jest.spyOn(GlBreakpointInstance, 'getBreakpointSize').mockReturnValue(bp);

          wrapper.vm.onResize();
          await wrapper.vm.$nextTick();

          expect(findGlPopover().attributes('disabled')).toBe(isDisabled);
        },
      );
    });

    describe('onShown', () => {
      it('dispatches tracking event', () => {
        findGlPopover().vm.$emit('shown');

        expectTracking(trackingEvents.popoverShown);
      });
    });
  });
});
