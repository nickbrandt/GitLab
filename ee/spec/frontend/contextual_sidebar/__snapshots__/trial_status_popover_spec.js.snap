// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrialStatusPopover component matches the snapshot 1`] = `
<gl-popover-stub
  boundary="viewport"
  cssclasses=""
  delay="[object Object]"
  placement="rightbottom"
  target="target-element-identifier"
  triggers="hover focus"
>
   
  <gl-sprintf-stub
    message="Your trial ends on %{boldStart}%{trialEndDate}%{boldEnd}. We hope you’re enjoying the features of GitLab %{planName}. To keep those features after your trial ends, you’ll need to buy a subscription. (You can also choose GitLab Premium if it meets your needs.)"
  />
   
  <div
    class="gl-mt-5"
  >
    <gl-button-stub
      block=""
      buttontextclasses=""
      category="primary"
      class="gl-mb-0"
      data-testid="upgradeBtn"
      href="transactions/new"
      icon=""
      size="small"
      variant="confirm"
    >
      <span
        class="gl-font-sm"
      >
        Upgrade Some Test Group to Ultimate
      </span>
    </gl-button-stub>
     
    <gl-button-stub
      block=""
      buttontextclasses=""
      category="secondary"
      class="gl-mb-0"
      data-testid="compareBtn"
      href="billing/path-for/group"
      icon=""
      size="small"
      title="Compare all plans"
      variant="confirm"
    >
      <span
        class="gl-font-sm"
      >
        Compare all plans
      </span>
    </gl-button-stub>
  </div>
</gl-popover-stub>
`;
